### DISCUSSION TITLE
Return type has to be explicitly typed

### DISCUSSION THREAD

**Discussion by vojta-horanek** [LemonAppDev/konsist#620] [Report a Problem]
### Tell us what version of Konsist you use.

0.12.1
### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

When a type for a function or a property is inferred, Konsist will not be able to detect it properly. Instead of the actual declaration return type, Konsist says it's null.

```kotlin
@Test
fun fails() {
    Konsist
        .scopeFromProject()
        .functions()
        .filter { it.name == "uniqueTestFunction" }
        .assert { it.returnType?.name == "String" }
}

fun uniqueTestFunction() = "Failing" 
```

This is particularly a problem when wanting to filter out functions which do return a value from those that return a `Unit`. For example when wanting to assert on all the `@Composable` functions that emit UI.

### Actual behaviour

The type is not detected properly and null is given by Konsist when asked about the return type.

### Expected behaviour

The return type returned should be detected properly even when it is not explicitly "typed".

**Comment by igorwojda** [comment]
Kotlin compiler which has a deeper understanding of the type system. Konsist analyzes text in Kotlin files, so if the type is not specified Konsist does not know it. Thats being said we should be able to determine that the return type is `Unit` just by looking at the declaration text itself (I will add a ticket to the backlog).
Also in the future, we may be able to determine this type using [K2 Compiler plugins](https://www.youtube.com/watch?v=Pl-89n9wDqo).

**Comment by igorwojda** [comment]
As a workaround, you can look for a "return" string in the declaration body.

```kotlin
val sut = getSnippetFile("declaration-has-no-return-type")
            .functions()
            .filter { it.name == "uniqueTestFunction" }
            .assert { it.text.contains("return") }
```

The logic for determining if the function returns a value is more complex (expression body vs. block body, lambdas in side function may have local or non-local returns...), but you should get the idea and write the required check by processing declaration `text`.

**Comment by igorwojda** [comment]
A new version v0.13.0 has been released. It contains the fix.
See https://github.com/LemonAppDev/konsist/releases/tag/v0.13.0

It adds

```Kotlin
Konsist
            .scopeFromProject()
            .functions()
            .assertTrue { it.returnType?.text == "Unit" || it.returnType?.text == null } 
```
